### ISSUE TITLE
`invalid-argument-type` false positive: generic type parameter resolved to `Unknown` after `or` expression

### ISSUE THREAD

**Issue by eitan-protego** [astral-sh/ty#3693]
### Summary

Generic type parameter lost (resolved to `Unknown`/`str`) in `or` expression with `None` fallback

## Description

When a generic function returns `dict[T, V] | None` and the result is used in an `or` expression with a dict literal fallback, ty loses the resolved generic type parameter. The left branch of the `or` shows `T` as `Unknown` (or widened to `str`) instead of the concrete `Literal[...]` type that was resolved from the call.

This blocks a pattern where a builder function returns `None` when there's nothing to build, and `or` provides a default:

```python
sort = build_sort_spec(params) or {"default_field": -1}
```

### Repro ([Playground](https://play.ty.dev/7ae91ffd-dd8d-44d5-ab98-f23ac95a0a94))

```python
from typing import Literal
from collections.abc import Mapping
from dataclasses import dataclass


@dataclass
class SortParams[F]:
    field: F
    direction: Literal["asc", "desc"] = "desc"


def build_sort_spec[T](
    sort_params: SortParams[T] | None,
) -> dict[T, Literal[1, -1]] | None:
    if not sort_params:
        return None
    return {sort_params.field: 1}


type Path = Literal["name", "age", "created"]


def use_sort(s: Mapping[Path, Literal[1, -1]]) -> None: ...


params: SortParams[Path] | None = None
sort = build_sort_spec(params) or {"name": -1}
# ty infers sort as: (dict[Unknown, Literal[1, -1]] & ~AlwaysFalsy) | dict[str, int]
# Expected:          dict[Path, Literal[1, -1]] | dict[Literal["name"], Literal[-1]]
use_sort(sort)
```

### Expected behavior

`build_sort_spec(params)` should resolve `T = Path`. After the `or` expression:
- Left branch: `dict[Path, Literal[1, -1]]` (the non-`None` case)
- Right branch: `dict[Literal["name"], Literal[-1]]` (the fallback literal)

Both should be assignable to `Mapping[Path, Literal[1, -1]]`.

### Actual behavior

```
error[invalid-argument-type]: Argument to function `use_sort` is incorrect:
  Expected `Mapping[Path, Literal[1, -1]]`,
  found `(dict[Unknown, Literal[1, -1]] & ~AlwaysFalsy) | dict[str, int]`
```

Two problems:
1. **Left branch**: `T` is lost — shown as `Unknown` instead of `Path`
2. **Right branch**: dict literal `{"name": -1}` is widened to `dict[str, int]` instead of `dict[Literal["name"], Literal[-1]]`

### Other type checkers

| Checker | Result |
|---|---|
| pyright 1.1.400 | pass |
| basedpyright 1.39.2 | pass |
| mypy 1.20.2 | pass |
| pyrefly 0.64.1 | pass |
| **ty 0.0.44** | **error** |

### Version

ty 0.0.44 (f5523e256 2026-06-04)

**Comment by sharkdp** [comment]
Thank you for reporting this.

I think the problem here is that we don't infer the right-hand-side of the binary `... or ...` expression with type context. When the intermediate `sort` variable is removed (i.e. if you pass `... or ...` into the `use_sort` call directly), everything works as expected.

So in order to support this... we would need to somehow use the return type of the left-hand-side of `build_sort_spec(params) or {"name": -1}` as type context for the right-hand side of the `or` (for the dict literal)?

Here's a smaller reproducer (which all other type checkers accept):
```py
from typing import Literal

type Key = Literal["foo", "bar"]

def f(x: dict[Key, int]):
    ...

def main(d: dict[Key, int] | None):
    x = d or {"foo": 0}
    f(x)
```
https://play.ty.dev/33753b06-7808-4f8b-bb26-f85dc8c03ec1

**Comment by carljm** [comment]
This is the same underlying issue (and requires the same fix) as #3574. I have a branch in-progress that fixes it.
